### DISCUSSION TITLE
Getting started with Docker and manticore Search

### DISCUSSION THREAD

**Discussion by nyluntu** [sovelluskontti/project-semantic-search#40] [General]
Hi,

ping @sarafarahabadi 

I tested how columnar storage works as a Docker container. The tutorial commands worked right away once columnar storage was enabled.

I suggest we continue using the Docker environment to reduce configuration requirements on our operating system.

Here are a few tutorials you should try to verify that Docker works on your machine first. After that, you can try setting up the Manticore container image.

# Docker Tutorials

- First, install Docker Desktop: https://www.docker.com/products/docker-desktop/
    - No account creation is needed.
- Main tutorial: https://docs.docker.com/get-started/
    - Get familiar with Docker concepts: https://docs.docker.com/get-started/docker-overview/
    - Complete parts 1 and 2. There’s no need to push images into a container registry. The important thing is verifying that 
    - Docker works on your machine and that you can view the example tutorial app in your browser: https://docs.docker.com/get-started/introduction/
- Additional concepts: https://docs.docker.com/get-started/docker-concepts/the-basics/what-is-a-container/
    - This section covers basics like building image layers and running containers. You only need to read through it, no testing required.

# Manticore Search

I’ll push the Docker Compose configuration to our repository to ensure it works for both of us. Docker Compose is an important concept to understand. I’ll provide more guidance on how I set things up on my machine, though it might be different for you, as I’m using macOS and haven’t tested everything on Windows recently.

Let’s have a follow-up meeting later this week to review this together.

Example from my machine showing how the search worked:

```
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 4
Server version: 6.3.6 593045790@24080214 (columnar 2.3.0 88a01c3@24052206) (secondary 2.3.0 88a01c3@24052206) (knn 2.3.0 88a01c3@24052206) git branch manticore-6.3.6...origin/manticore-6.3.6

Copyright (c) 2000, 2020, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> show tables;
Empty set (0.00 sec)

mysql> show database;
ERROR 1064 (42000): P01: syntax error, unexpected identifier, expecting VARIABLES near 'database'
mysql> use Manticore;
Database changed
mysql> show tables;
Empty set (0.00 sec)

mysql> show databases;
+-----------+
| Databases |
+-----------+
| Manticore |
+-----------+
1 row in set (0.00 sec)

mysql> create table test ( title text, image_vector float_vector knn_type='hnsw' knn_dims='4' hnsw_similarity='l2' );
Query OK, 0 rows affected (0.00 sec)

mysql> describe table test;
ERROR 1064 (42000): P01: syntax error, unexpected identifier, expecting $end near 'test'
mysql> describe test;
+--------------+--------------+----------------+
| Field        | Type         | Properties     |
+--------------+--------------+----------------+
| id           | bigint       |                |
| title        | text         | indexed stored |
| image_vector | float_vector | knn            |
+--------------+--------------+----------------+
3 rows in set (0.00 sec)

mysql> insert into test values ( 1, 'yellow bag', (0.653448,0.192478,0.017971,0.339821) ), ( 2, 'white bag', (-0.148894,0.748278,0.091892,-0.095406) );
Query OK, 2 rows affected (0.02 sec)

mysql> select id, knn_dist() from test where knn ( image_vector, 5, (0.286569,-0.031816,0.066684,0.032926), 2000 );
+------+------------+
| id   | knn_dist() |
+------+------------+
|    1 | 0.28146550 |
|    2 | 0.81527930 |
+------+------------+
2 rows in set (0.00 sec)
--- 2 out of 2 results in 0ms ---

mysql> 
```

**Comment by sarafarahabadi** [comment]
Hello, 

Thank you so much for the information. I will make the necessary changes to use Docker and make the semantic search work. 
I am available Thursday for the meeting at any time. 

Also, I entered my information in the OneDrive file and submitted my tax card. 

Br, 
Sara

**Reply by nyluntu** [reply]
I uploaded a short video about Docker Compose to the shared folder. In it, I cover the basics of how it works once Docker is installed and how to use a Docker Compose file. When everything is set up correctly, it only takes a single command to launch the container. Let's check on Thursday to see if everything works on your machine as well.

It's probably good to check some overview about the Docker https://www.youtube.com/watch?v=pg19Z8LL06w

Our environment is very simple and doesn’t require any custom containers at the moment. If we need a web application in the future, then custom containers can be used and configured as needed.

**Reply by sarafarahabadi** [reply]
Hi, 
Thank you so much for the video. It was a big help. I ran the test and it works on my system as well now. I will now work to upload the data into the table there and make the search work. 
Br, 
Sara
